jetty-util/src/main/java/org/eclipse/jetty/util/thread/strategy/AdaptiveExecutionStrategy.java

```diff
@@ -0,0 +1,465 @@
+    }
+
+    @Override
+    public void produce()
+    {
+        tryProduce(false);
+    }
+
+    private void tryProduce(boolean wasPending)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} tryProduce {}", this, wasPending);
+
+        try (AutoLock l = _lock.lock())
+        {
+            if (wasPending)
+                _pending = false;
+
+            switch (_state)
+            {
+                case IDLE:
+                    // Enter PRODUCING
+                    _state = State.PRODUCING;
+                    break;
+
+                case PRODUCING:
+                    // Keep other Thread producing
+                    _state = State.REPRODUCING;
+                    return;
+
+                default:
+                    return;
+            }
+        }
+
+        boolean nonBlocking = Invocable.isNonBlockingInvocation();
+
+        while (isRunning())
+        {
+            try
+            {
+                if (doProduce(nonBlocking))
+                    continue;
+                return;
+            }
+            catch (Throwable th)
+            {
+                LOG.warn("Unable to produce", th);
+            }
+        }
+    }
+
+    private boolean doProduce(boolean nonBlocking)
+    {
+        Runnable task = produceTask();
+
+        if (task == null)
+        {
+            try (AutoLock l = _lock.lock())
+            {
+                // Could another task just have been queued with a produce call?
+                switch (_state)
+                {
+                    case PRODUCING:
+                        _state = State.IDLE;
+                        return false;
+
+                    case REPRODUCING:
+                        _state = State.PRODUCING;
+                        return true;
+
+                    default:
+                        throw new IllegalStateException(toStringLocked());
+                }
+            }
+        }
+
+        Mode mode;
+        Invocable.InvocationType taskType = Invocable.getInvocationType(task);
+        if (taskType == Invocable.InvocationType.NON_BLOCKING)
+        {
+            mode = Mode.PRODUCE_CONSUME;
+        }
+        else if (!nonBlocking)
+        {
+            try (AutoLock l = _lock.lock())
+            {
+                if (_pending || _tryExecutor.tryExecute(this))
+                {
+                    _pending = true;
+                    _state = State.IDLE;
+                    mode = Mode.EXECUTE_PRODUCE_CONSUME;
+                }
+                else
+                {
+                    mode = taskType == Invocable.InvocationType.EITHER
+                        ? Mode.PRODUCE_INVOKE_CONSUME
+                        : Mode.PRODUCE_EXECUTE_CONSUME;
+                }
+            }
+        }
+        else
+        {
+            mode = taskType == Invocable.InvocationType.EITHER
+                ? Mode.PRODUCE_INVOKE_CONSUME
+                : Mode.PRODUCE_EXECUTE_CONSUME;
+        }
```

Comment:
I think this block should be reverted back to its original code, for 2 reasons:
1. This is a 3-way if/else with the first `if` having a different condition than the second if/else. The previous version was clearer in that it only had a if/else on the thread blocking/non-blocking and was so much easier to read.
2. The previous version had a lot of good comments that have been all stripped out but they are very important to keep.

The cost is exactly the same, so I prefer the previous version because it allows better reasoning by splitting all the cases, rather than this version that coalesces multiple cases.

Reply:
which previous version?  The version prior to this PR or the version prior to the last commit?

The problem with the version prior to the last commit was that it didn't lock.  To add locking means that it either needs to take the lock when it doesn't need to, or some code needs to be reproduced, or we need to add a null check.  

Let me try some of those alternatives...

Reply:
Oh you must have meant the version before this PR.  Nah!!!! the comments were clear, but the code was impenatrable and too long.  It took me some time to extract the simple description of the algorithm for the javadoc.    I have pushed another version of the refactored algorithm, this time with good comments and a little "hack" on the lock to avoid code duplication and/or null checks.